domain/application/state/machine_placement.go

```diff
@@ -0,0 +1,235 @@
+		// For example, lxd.
+		scope := directive.Container.String()
+		childNetNode, err := st.insertChildMachineForContainerPlacement(ctx, tx, machineUUID, machineName, scope)
+		if err != nil {
+			return "", errors.Errorf("inserting child machine for container placement: %w", err)
+		}
+		return childNetNode, nil
+
+	case placement.PlacementTypeProvider:
+		// The placement is handled by the provider, so we need to create a
+		// machine for the net node and then insert the provider placement
+		// for the machine.
+		machineName, err := st.nextMachineSequence(ctx, tx)
+		if err != nil {
+			return "", errors.Capture(err)
+		}
+
+		machine, netNode, err := st.insertMachineForNetNode(ctx, tx, machineName)
+		if err != nil {
+			return "", errors.Capture(err)
+		}
+		if err := st.insertMachineProviderPlacement(ctx, tx, machine, directive.Directive); err != nil {
+			return "", errors.Errorf("inserting machine provider placement: %w", err)
+		}
+		return netNode, nil
+
+	default:
+		return "", errors.Errorf("invalid placement type: %v", directive.Type)
+	}
+}
+
+func (st *State) getMachineNetNodeUUIDFromName(ctx context.Context, tx *sqlair.TX, name machine.Name) (string, error) {
+	machine := machineName{Name: name}
+	query := `
+SELECT &machineName.net_node_uuid
+FROM machine
+WHERE name = $machineName.name
+`
+	stmt, err := st.Prepare(query, machine)
+	if err != nil {
+		return "", errors.Capture(err)
+	}
+	err = tx.Query(ctx, stmt, machine).Get(&machine)
+	if errors.Is(err, sqlair.ErrNoRows) {
+		return "", errors.Errorf("machine %q not found", name).
+			Add(applicationerrors.MachineNotFound)
+	} else if err != nil {
+		return "", errors.Errorf("querying machine %q: %w", name, err)
+	}
+	return machine.NetNodeUUID, nil
+}
+
+func (st *State) insertNetNode(ctx context.Context, tx *sqlair.TX) (string, error) {
+	uuid, err := uuid.NewUUID()
+	if err != nil {
+		return "", errors.Capture(err)
+	}
+
+	netNodeUUID := netNodeUUID{NetNodeUUID: uuid.String()}
+
+	createNode := `INSERT INTO net_node (uuid) VALUES ($netNodeUUID.*)`
+	createNodeStmt, err := st.Prepare(createNode, netNodeUUID)
+	if err != nil {
+		return "", errors.Capture(err)
+	}
+
+	if err := tx.Query(ctx, createNodeStmt, netNodeUUID).Run(); err != nil {
+		return "", errors.Errorf("creating net node for machine: %w", err)
+	}
+
+	return netNodeUUID.NetNodeUUID, nil
+}
+
+func (st *State) insertMachineForNetNode(ctx context.Context, tx *sqlair.TX, machineName machine.Name) (machine.UUID, string, error) {
+	netNodeUUID, err := st.insertNetNode(ctx, tx)
+	if err != nil {
+		return "", "", errors.Capture(err)
+	}
+
+	machineUUID, err := machine.NewUUID()
+	if err != nil {
+		return "", "", errors.Capture(err)
+	}
+
+	m := createMachine{
+		MachineUUID: machineUUID,
+		NetNodeUUID: netNodeUUID,
+		Name:        machineName,
+		LifeID:      life.Alive,
+	}
+
+	createMachineQuery := `
+INSERT INTO machine (uuid, net_node_uuid, name, life_id)
+VALUES ($createMachine.*);
```

Comment:
I'm afraid this will bite us soon. Reading from tables that don't belong to the domain seems OK, but inserting? This concerns the machine domain so it should maybe live there?

Reply:
Machines will be moving to application domain. The application domain is going to be renamed deploy.